superset/assets/src/query/buildDatasource.ts

```diff
@@ -0,0 +1,35 @@
+import FormData from './FormData';
+
+enum DatasourceType {
+  Table = 'table',
+  Druid = 'druid',
+}
+
+export interface DatasourceKey {
+  id: number;
+  type: DatasourceType;
+}
+
+// Declaration merging with the interface above. No need to redeclare id and type.
+export class DatasourceKey {
```

Comment:
One little tweak here. Let's `export default class DatasourceKey`, remove `buildDatasource` and make `buildQueryContext` call `new DatasourceKey(formData.datasource).toObject()` manually. 

I think that will make this file really clean (decoupled from `Formdata`).